### ISSUE TITLE
spring-boot-starter-websocket breaks after adding org.springframework.integration:spring-integration-jmx to build.gradle

### ISSUE THREAD

**Issue by SenorLlama** [spring-projects/spring-boot#2441]
I'm trying to use websockets while also using jmx to monitor some integration beans.

Application works perfectly fine with websockets in build path. Snippet of build.gradle:

```gradle
def spring = [
        'org.springframework.boot:spring-boot-starter-web',
        'org.springframework.boot:spring-boot-starter-security',
        'org.springframework.boot:spring-boot-starter-actuator',
        'org.springframework.boot:spring-boot-starter-jersey',
        'org.springframework.boot:spring-boot-starter-integration',
        'org.springframework.boot:spring-boot-starter-amqp',
        'org.springframework.boot:spring-boot-starter-data-jpa',
        'org.springframework.boot:spring-boot-starter-data-mongodb',
        'org.springframework:spring-context-support',
        'org.springframework:spring-orm',
        'org.springframework:spring-oxm',
        'org.springframework.integration:spring-integration-xml',
        'org.springframework.integration:spring-integration-amqp',
        "org.springframework.boot:spring-boot-starter-websocket"
]

dependencies {
    compile spring, ...
    testCompile springTest
}
```

however, when I add jmx to the build path 
```gradle
def spring = [
        'org.springframework.boot:spring-boot-starter-web',
        'org.springframework.boot:spring-boot-starter-security',
        'org.springframework.boot:spring-boot-starter-actuator',
        'org.springframework.boot:spring-boot-starter-jersey',
        'org.springframework.boot:spring-boot-starter-integration',
        'org.springframework.boot:spring-boot-starter-amqp',
        'org.springframework.boot:spring-boot-starter-data-jpa',
        'org.springframework.boot:spring-boot-starter-data-mongodb',
        'org.springframework:spring-context-support',
        'org.springframework:spring-orm',
        'org.springframework:spring-oxm',
        'org.springframework.integration:spring-integration-xml',
        'org.springframework.integration:spring-integration-amqp',
        "org.springframework.boot:spring-boot-starter-websocket",
        'org.springframework.integration:spring-integration-jmx'
]
```

I get the following error on `gradle clean bootRun`
```
Failed to instantiate [org.springframework.messaging.support.AbstractSubscribableChannel]: Factory method 'brokerChannel' threw exception; nested exception is java.lang.ClassCastException: com.sun.proxy.$Proxy104 cannot be cast to org.springframework.messaging.support.AbstractSubscribableChannel
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:189)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:588)
	... 52 common frames omitted
Caused by: java.lang.ClassCastException: com.sun.proxy.$Proxy104 cannot be cast to org.springframework.messaging.support.AbstractSubscribableChannel
	at org.springframework.web.socket.config.annotation.DelegatingWebSocketMessageBrokerConfiguration$$EnhancerBySpringCGLIB$$b9022f77.clientInboundChannel(<generated>)
	at org.springframework.messaging.simp.config.AbstractMessageBrokerConfiguration.getBrokerRegistry(AbstractMessageBrokerConfiguration.java:212)
	at org.springframework.messaging.simp.config.AbstractMessageBrokerConfiguration.brokerChannel(AbstractMessageBrokerConfiguration.java:180)
	at org.springframework.web.socket.config.annotation.DelegatingWebSocketMessageBrokerConfiguration$$EnhancerBySpringCGLIB$$b9022f77.CGLIB$brokerChannel$26(<generated>)
	at org.springframework.web.socket.config.annotation.DelegatingWebSocketMessageBrokerConfiguration$$EnhancerBySpringCGLIB$$b9022f77$$FastClassBySpringCGLIB$$845f2078.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:228)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:309)
	at org.springframework.web.socket.config.annotation.DelegatingWebSocketMessageBrokerConfiguration$$EnhancerBySpringCGLIB$$b9022f77.brokerChannel(<generated>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMetssorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:162)
	... 53 common frames omitted
```

and app does not start.

Application.java is :
```java
@Configuration
@ComponentScan
@EnableJpaRepositories
@EnableMongoRepositories
@EnableEntityLinks
@EnableSwagger
@EnableAutoConfiguration
@ImportResource({
        "securityContext.xml"
})
public class Application {
    public static void main(String[] args) {
        SpringApplication.run(Application.class, args);
    }
}
```

**Comment by philwebb** [comment]
I'm having some trouble replicating the issue. Are you able to provide a [sample application](https://github.com/spring-projects/spring-boot-issues)?

My guess is that something is trying to apply AOP advice to `AbstractSubscribableChannel` using standard JDK Proxies which only work with interfaces. Without a sample app to reproduce the issue it's hard to track down what.

**Comment by SenorLlama** [comment]
Hey @philwebb ,

I created a barebones app here:

https://github.com/SenorLlama/spring-boot-sample/tree/master

The initial commit is a working sample of this tutorial:

https://spring.io/guides/gs/messaging-stomp-websocket/

The commit following that adds 'org.springframework.integration:spring-integration-jmx' to the build and results in the same error above.

Thanks for the help

**Comment by SenorLlama** [comment]
Noticed the issues repo after the above commit. I added the same sample in that repo. The pull request is here:

https://github.com/spring-projects/spring-boot-issues/pull/30

thanks

**Comment by philwebb** [comment]
This is a pretty interesting issue. Here's what I think is happening:

The `IntegrationMBeanExporter` class is trying to apply advice to the `send` and `receive` methods of all `MessageChannel` implementations.

The `AbstractMessageBrokerConfiguration.clientInboundChannel()` is a `MessageChannel` so the advice is applied to it when the bean is first created (when clientInboundChannel() is called from `getBrokerRegistry()`.

The `IntegrationMBeanExporter.applyAdvice()` method is called to apply the advice, it creates a `ProxyFactory` but notice the following comment:

```java
/**
* N.B. it's not a good idea to use proxyFactory.setProxyTargetClass(true) here because it forces all
* the integration components to be cglib proxyable (i.e. have a default constructor etc.), which they
* are not in general (usually for good reason).
*/
```

This means that the AOP framework can *only* proxy interfaces. The advice is applied and the resulting object is a valid `SubscribableChannel` but it is no longer a `AbstractSubscribableChannel`. Then the `@Configuration` method result is returned a `ClassCastException` is throw. 

I've simplified the repro project a little so that it can run without a local database: https://github.com/philwebb/boot-gh-2441

**Comment by philwebb** [comment]
@rstoyanchev  @garyrussell @artembilan I'm not sure what the best fix for this issue is. I don't think that there is much we can do in Boot. Rossen, could we change the `AbstractMessageBrokerConfiguration.clientInboundChannel()` method to return a `SubscribableChannel` rather than a `AbstractSubscribableChannel`?

**Comment by artembilan** [comment]
@philwebb , you are right. Even if those WebSocket support channels (`clientInboundChannel`, `clientOutboundChannel`, `brokerChannel`) aren't populated from the Spring Integration and even if we can somehow overcome JMX exposing for those channels from Spring Integration, there is no guarantees that those beans won't be proxied through the other way.
BTW, the same issue we see and with `AbstractMessageBrokerConfiguration.simpAnnotationMethodMessageHandler` and other `MessageHandler` beans which aren't presented as an interface return type.
Yes, there is nothing to do with Boot. It's an issue of Spring Messaging and Spring WebSockets together with Spring Integration JMX support.
That's how we can fix it already now with backport. Seems for me there is a rule in Spring Framework: avoid target classes as much as possible. Extract interfaces if you can.

For the future we maybe should think how to exclude those Messaging beans which aren't for the Spring Integration layer. 
However I think it isn't a case: maybe would be better to move `IntegrationMBeanExporter` to the Spring Messaging since `MessageChannel` and `MessageHandler` are located there and not only Spring Websockets and Spring Integration use them. E.g. Spring Cloud Bus.

@rstoyanchev , your turn!

**Comment by artembilan** [comment]
Speak of the devil and he is sure to appear:
http://stackoverflow.com/questions/28316008/spring-xd-spring-integration-dsl-based-aws-s3-sink-module-error
:smile:

**Comment by artembilan** [comment]
@SenorLlama , BTW you can workaround this issue with `spring.jmx.enabled=false`, if you aren't interested in the JMX exposing.
However if you need JMX anyway you can switch off `IntegrationMBeanExporter` with the:
````java
@EnableIntegrationMBeanExport(defaultDomain = "${spring.jmx.default_domain:}", server = "${spring.jmx.server:mbeanServer}", managedComponents="")
````
on one of your `@Configuration` to suppress any component name to be matched for the exposing.

**Comment by SenorLlama** [comment]
@artembilan 
I will be using spring integration and do want MBeanExporter to manage components and shut them down gracefully should something arise so don't think I can resort to using ```@EnableIntegrationMBeanExport``` . Moreover, using the above causes an error on startup as managedComponents cannot be empty.

```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'integrationMbeanExporter' defined in class org.springframework.integration.jmx.config.IntegrationMBeanExportConfiguration: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.integration.monitor.IntegrationMBeanExporter]: Factory method 'mbeanExporter' threw exception; nested exception is java.lang.IllegalArgumentException: componentNamePatterns must not be empty
```

**Comment by artembilan** [comment]
OK, @SenorLlama , my bad. Yes, `componentNamePatterns` can't be empty. So `0` is a good candidate to suppress any bean to be exposed, because I'm not sure that you provide such a bean name for some component :smile: .
Another case to overcome that issue with Spring Websockets, just specify `componentNamePatterns` with names (patterns) for those Spring Integration components in which you are interested in.

**Comment by SenorLlama** [comment]
Hey @artembilan. np lol.

I tried the suggestion. However, regardless of what patterns I provided in ```managedComponents``` even the ```0```, the same error comes up. 

On a different note, I tried playing around with that annotation and when I use it the monitor mbean disappears from jconsole:
![screen shot 2015-02-04 at 11 36 00 am](https://cloud.githubusercontent.com/assets/2054628/6048002/2e9b122a-ac62-11e4-9a8a-00a1e96478de.png)
vs:
![screen shot 2015-02-04 at 11 32 19 am](https://cloud.githubusercontent.com/assets/2054628/6048008/36f78caa-ac62-11e4-91f3-6248c6bcaf01.png)

The latter is just a side note though. The error seems to be coming from spring messaging proxying. 

if I set ```spring.jmx.enabled=false``` then the app does start but I do need jmx

**Comment by rstoyanchev** [comment]
@artembilan, @philwebb, a quick check shows that switching to SubscribableChannel doesn't cause any test failures, just one test compilation issue that's fixable. I'd have to do a little more checking. I know that with Java config the return type from the @Bean method determines the type of the bean which is important for detecting lifecycle interfaces. That's why switching from SimpAnnotationMethodMessageHandler (implements SmartLifecycle) to MessageHandler probably wouldn't work.

**Comment by artembilan** [comment]
Doh, @SenorLlama . Yes, I missed some code from the `IntegrationMBeanExporter`:
```
if (bean instanceof MessageHandler) {
 ......
	SimpleMessageHandlerMetrics monitor = new SimpleMessageHandlerMetrics((MessageHandler) bean);
	Object advised = applyHandlerInterceptor(bean, monitor, beanClassLoader);
	handlers.add(monitor);
	bean = advised;
}
```
So, we always proxy any beans which are instance of `MessageHandler`, `MessageSource` or `MessageChannel`. And only decide to expose them to the JMX, if they much that `componentNamePatterns`. 

So, it is a critical bug and ... I don't see a workaround. First of all I wasn't able to find how to override `@Bean`s for `DelegatingWebSocketMessageBrokerConfiguration`. And even can't do that with `@ImportResource("websocket-config.xml")`, because `WebSocketMessageBrokerStats` requires `SubProtocolWebSocketHandler` and `StompBrokerRelayMessageHandler` which are proxy too. 
But even if I disable `WebSocketMessageBrokerStats`:
````xml
<bean id="webSocketMessageBrokerStats" class="java.lang.Object"/>
````
I end up with a naming issue:
> Caused by: javax.management.MalformedObjectNameException: Invalid character `:' in value
	at javax.management.ObjectName.parseValue(ObjectName.java:974)

That's because `AbstractBrokerMessageHandler` isn't a `NamedComponent` (from Spring Integration) and we call `toString()` for the MBean name. And that value isn't valid for `javax.management.ObjectName`.

So, the issue is more broader and we really should have some internal discussion how to live with that.

**Comment by artembilan** [comment]
@rstoyanchev , just tested that:
````java
@Bean
public MessageHandler userDestinationMessageHandler() {
	return new UserDestinationMessageHandler(new DirectChannel(), new DirectChannel(), new
					DefaultUserDestinationResolver(new DefaultUserSessionRegistry()));
}
````
Works fine: the `AbstractApplicationContext` calls `	getLifecycleProcessor().onRefresh();` and my break point in the `UserDestinationMessageHandler.start()` is reached.

So, we should collect all these issues in the separate thread and share thoughts what to do.

cc. @garyrussell

**Comment by artembilan** [comment]
BTW, there is nothing to do with Boot. We really must move somewhere other and stop bother Boot guys :smile:

**Comment by garyrussell** [comment]
I would rather not move the `IMBE` to `spring-messaging` at this time. I have long wanted to remove the use of proxies for JMX support of integration component and perhaps this is another impetus to do so.

Of course that won't solve the general issue of proxied channels (for other reasons) where the APIs expect an abstract channel.

**Comment by artembilan** [comment]
We are going to get rid of that proxying https://jira.spring.io/browse/INT-3627.
Hence any other projects who use `MessageChannel`, `MessageHandler` won't be affected.
@SenorLlama , I found one more workaround how to disable SI JMX export:
````java
@Bean
public IntegrationMBeanExporter integrationMbeanExporter() {
	return new IntegrationMBeanExporter() {

		@Override
		public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
			return bean;
		}

	};
}
````

But I understand that it isn't your case, because you want to manage SI endpoints through the JMX...

**Comment by philwebb** [comment]
Closing this one as it will be addressed in the 4.2 Spring Integration release.

**Comment by artembilan** [comment]
Yes, we already have a fix in the SI-4.2-BUILD-SNAPSHOT, but we should bear in mind that it will work only with Boot 1.3, because SI-4.2 is based on the SF-4.2.
A brief description what we did:
* Instead of proxying MessageChannels/MessageHandlers/MessageSources we moved metrics and statistics to the base abstract classes. Hence SI components support them directly without proxy interceptors.
* Only SI components are enhanced to do those metrics, because they extend exactly SI classes.
* The `IntegrationMBeanExporter` gets deal only with those components which are marked with `@IntegrationManagedResource` (or their superclasses/interfaces).
* Having such a fix we don't know about any other `MessageChannel`/`MessageHandler` implementation from the Spring WebSockets or other projects which are based on the Spring Messaging.

Note: if `MessageChannel`/`MessageHandler` doesn't extend SI `AbstractMessageChannel`/`AbstractMessageHandler` there is no reason to mark your own classes with `@IntegrationManagedResource`, because they won't have a value without SI metrics. Hence you can simply use standard `MBeanExporter` with standard `@ManagedResource`.
